Genmod rankscore analysis report

torbjorgen opened the pull request:
This PR contains documentation presenting the results for testing RankScore and RankScoreNormalized integrity and performance metrics of Genmod with respect to MutAcc variants.

Genmod source code used for testing is in this PR: https://github.com/Clinical-Genomics/genmod/pull/108

torbjorgen committed:
RankScore test: Validation report

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen commented:
Hi @jemten  and @henrikstranneheim 

This PR is blocked because I'm unable to reproduce the dataset with new, patched genmod version from PR https://github.com/Clinical-Genomics/genmod/pull/108.

I need a new dataset that's been produced with the above PR to rerun the tests for structural variants, to generate the plots in this PR.

Can you please help me out here? Much appreciated.

jemten commented:
I can help out with that. I'll drop by your desk and you can show me what's needed

torbjorgen committed:
RankScore test: Update wrapper shellscript with new data paths

torbjorgen committed:
RankScore test: Disable dataset comparison for SV

The reference dataset contains SNPs, not SVs, so the comparison
to reference data set cannot be done for the SV case.

Allow None path as reference.

Adjust the evaluation bash script to fit to this case.

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen committed:
RankScore test: Introduce variant UID for dataset comparision, low bound check

It was observed that in comparing reference and test set,
variants rank scores got mixed up due to identical
VCF.ID values, causing the RankScore test to fail.

As a solution, introduce a new UID based on VCF fields:
 * POS
 * ID
 * ALT
 * REF

so that there can be no ambiguity in variant sorting in
the dataframe.

Furthermore, allow RankScores to differ in between
reference set and test set, due to newly introduced
RankScore capping behavior in Genmod.
See more information in patch itself.

Signed-off-by: Tor Björgen <tor.bjorgen@scilifelab.se>

torbjorgen commented:
Thanks @jemten for supplying the data!! :+1:

henrikstranneheim commented on src/rdds/exploration_rankscore/assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,46 @@
+    ids = df.index.values
+    positions = df.pos.values
+    alts = df.alt.values
+    refs = df.ref.values
```
Add type hint

henrikstranneheim commented on src/rdds/exploration_rankscore/assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,46 @@
+                               id: bytes,
+                               alt: bytes,
+                               ref: bytes) -> bytes:
+    """
```
```suggestion
def assemble_unique_variant_id(chromosome_pos: int,
                               variant_id: bytes,
                               alt_allele: bytes,
                               ref_allele: bytes) -> bytes:
    """
```
I know this, but to someone new, but I think it  might be good for someone not familiar with the VCF format to be more explicit

henrikstranneheim commented on src/rdds/exploration_rankscore/assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,46 @@
+        unique_id = assemble_unique_variant_id(pos=pos, id=id, alt=alt, ref=ref)
+        unique_ids.append(unique_id)
+    unique_ids_df = pd.DataFrame(data={'UID': unique_ids}, index=ids)
+    df = pd.concat((df, unique_ids_df), axis=1)
```
Is there more mnemonic name than the type of container?

henrikstranneheim commented on src/tests/exploration_rankscore/test_assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,38 @@
+    # GIVEN a variant with POS, ID, ALT, and REF metadata
+    new_id = assemble_unique_variant_id(pos=1000, id=b'10_1032245_CA_C', alt=b'C', ref=b'CA')
+    # WHEN assembling a new UID, THEN expect it to be correct
+    assert new_id == b'1000-10_1032245_CA_C-alt-C-ref-CA'
```
```suggestion
    # WHEN assembling a new UID
    new_id = assemble_unique_variant_id(pos=1000, id=b'10_1032245_CA_C', alt=b'C', ref=b'CA')

   # THEN return a binary string
    assert new_id == b'1000-10_1032245_CA_C-alt-C-ref-CA'
```

henrikstranneheim commented on src/tests/exploration_rankscore/test_assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,38 @@
+    # GIVEN a dataframe with some dummy variants
+    df = pd.DataFrame(data={'pos': [0, 1], 'alt': [b'C', b'A'], 'ref': [b'A', b'C']}, index=[b'id0', b'id0'])
+    df = dataframe_assemble_unique_variant_id(df=df)
+    # WHEN assembling, reindexing the DF based on new UID, THEN expect the UIDs to be correct
```
```suggestion
    # WHEN assembling, reindexing the DF based on new UID
    df = dataframe_assemble_unique_variant_id(df=df)

    # THEN  return data frame with 2 UIDs 
```

henrikstranneheim commented on src/tests/exploration_rankscore/test_assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,38 @@
+    df = pd.DataFrame(data={'pos': [0, 0], 'alt': [b'C', b'C'], 'ref': [b'A', b'A']}, index=[b'id0', b'id0'])
+    # WHEN reindexing, THEN expect this to be caught
+    with pt.raises(ValueError):
+        dataframe_assemble_unique_variant_id(df=df)
```
```suggestion
    # WHEN reindexing
    with pt.raises(ValueError):
        dataframe_assemble_unique_variant_id(df=df)
        
       # THEN an error should be raised
```

henrikstranneheim commented on src/tests/exploration_rankscore/test_assemble_unique_variant_id.py:
```diff
@@ -0,0 +1,38 @@
+import pandas as pd
+import pytest as pt
```
```suggestion
import pandas
import pytest
```

I think this improves readability and reduces cognitive load

henrikstranneheim reviewed (approved):
Awesome 👍 . Looks like they replicate beautifully 🎉

torbjorgen commented:
* Updated according to @henrikstranneheim suggestions.
